### PR TITLE
Improve conformance check

### LINKED ISSUE #241: Improve conformance check: Allow function subtypes
## Context

Currently conformance requires function implementations to be *equal*, including argument labels.
However, the parameters could be supertypes, and the return type could be a subtype:

https://github.com/onflow/cadence/blob/90cf100822e73355e95499693adeee6ee373d6c7/runtime/sema/check_composite_declaration.go#L885

For example, the following program should be valid, but is not at the moment:

```swift
pub resource interface RI {}

pub resource R: RI {}

pub struct interface SI {
    pub fun get(): @{RI}
}

pub struct S: SI {
    pub fun get(): @R {
        return <- create R()
    }
}
```

## Definition of Done

- [x] Sema: Improve conformance check for functions
- [x] Tests

### PR DESCRIPTION
**Pull request by turbolent**
Closes #241 

Make return types covariant, i.e. allow subtypes for return types of functions to satisfy interface conformance.

### CODE REVIEW

File: runtime/tests/checker/conformance_test.go [hunk 1]
```diff
@@ -198,38 +261,156 @@ func TestCheckTypeRequirementConformance(t *testing.T) {
...
+              fun get(): @R {
+                  return <- create R()
+              }
+          }
+        `)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("invalid, return type is supertype", func(t *testing.T) {
```
Comment (benjaminkvm) [inline runtime/tests/checker/conformance_test.go:296]: 👍

**Comment by benjaminkvm** [comment]
If this isn't in the Cadence documentation, would be really great to add it btw. One of those nuanced things that would be great to have written down.

**Comment by turbolent** [comment]
Yeah, we should maybe document it. 

When I implemented this, I checked several other languages, Swift, Rust, Scala, Kotlin, Java, C#, etc. and their behaviour is all different (some treat parameters and return type invariant, some allow covariant return types, etc.) – and it wasn't ever documented anywhere, i.e. I had to try it out.

**Comment by turbolent** [comment]
@benjaminkvm Added this to #237

**Pull request merged**